README.md
# Face Mask Detection using OpenVINO

| Details            |              |
|-----------------------|---------------|
| Programming Language: |  [![Python 3.7](https://img.shields.io/badge/python-3.7-blue.svg)](https://www.python.org/downloads/release/python-370/) |
| Intel OpenVINO ToolKit: |[![OpenVINO 2020.2](https://img.shields.io/badge/openvino-2020.2-blue.svg)](https://software.intel.com/content/www/us/en/develop/tools/openvino-toolkit/choose-download.html)|
| Docker (Ubuntu OpenVINO pre-installed): | [mmphego/intel-openvino](https://hub.docker.com/r/mmphego/intel-openvino)|
| Hardware Used: | Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz |
| Device: | CPU |
| Blog Post | [![blogpost](https://img.shields.io/badge/BlogPost-Link-brightgreen)](https://blog.mphomphego.co.za/blog/2020/06/02/Face-Mask-Detection-using-Intel-OpenVINO-and-OpenCV.html) |
| Visitors | ![](https://visitor-badge.laobi.icu/badge?page_id=mmphego.face_mask_detection_openvino)|


Face Mask Detection application uses Deep Learning/Machine Learning to recognize if a user is not wearing a mask and issues an alert.

By utilizing pre-trained models and Intel OpenVINO toolkit with OpenCV. This enables us to use the async API which can improve overall frame-rate of the application, rather than wait for inference to complete, the application can continue operating on the host while accelerator is busy.

This application executes 2 parallel infer requests for the Face Mask Detection and Face Detection networks that run simultaneously.

Using a set of the following pre-trained models:
- [face-detection-adas-0001](https://docs.openvinotoolkit.org/latest/_models_intel_face_detection_adas_0001_description_face_detection_adas_0001.html), which is a primary detection network for finding faces.
- face-mask-detection, which is a pretrained model for detecting a mask.

This application can be improved and then integrated with CCTV or other types cameras to detect and identify people without masks in public areas such as shopping centres and etc. This the ever increasing COVID-19 cases world-wide these application could be useful in controlling the spread of the virus.

![Screenshot from 2020-06-01 22-21-47](https://user-images.githubusercontent.com/7910856/83451683-a8d71780-a457-11ea-8eae-185725fefcc9.png)

**What is OpenVino?**

OpenVino (OpenVisual Inferencing and Neural Network Optimization) is toolkit to develop Deep Learning Application especially for Computer Vision by Intel. OpenVino Enables deep learning inference at the edge and supports heterogeneous execution across computer vision accelerators—CPU, GPU, Intel® Movidius™ Neural Compute Stick, and FPGA—using a common API. [read more](https://docs.openvinotoolkit.org/)

You might also be interested on reading about [AI At The Edge - An Introduction To Intel OpenVINO Toolkit.](https://blog.mphomphego.co.za/blog/2020/05/25/AI-at-the-Edge-An-introduction-to-Intel-OpenVINO-Toolkit.html)

## Support
If you have found this useful, please donate by clicking on the image below:

[![image](https://user-images.githubusercontent.com/7910856/88235803-e4ce7200-cc7b-11ea-8218-c3c04810052c.png)](https://paypal.me/mmphego)

### Tutorial
#### Blog Post

I have created a detailed blog post on the implementation: https://blog.mphomphego.co.za/blog/2020/06/02/Face-Mask-Detection-using-Intel-OpenVINO-and-OpenCV.html

#### YouTube Tutorial

The first of many...

[![Watch the video](https://user-images.githubusercontent.com/7910856/88237923-a8514500-cc80-11ea-9cc8-0692eb0c4d6e.gif)](https://www.youtube.com/watch?v=6r6foGbCHQ0)

## Hardware Requirement

- Minimum Intel Gen 6 processors


## Installation

- Download the docker images with a pre-installed version of OpenVINO 2020.2
```bash
docker pull mmphego/intel-openvino
```

- Download the facemask detection model.
```bash
wget https://github.com/didi/maskdetection/raw/master/model/face_mask.caffemodel
wget https://raw.githubusercontent.com/didi/maskdetection/master/model/deploy.prototxt
```

- Convert model to OpenVINO's `Intermediate Representations` (IR) using the `Model Optimizer`, which will produce `.xml` and `.bin` files.
```bash
docker run --rm -ti \
--volume "$PWD":/app \
--env DISPLAY=$DISPLAY \
--volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
mmphego/intel-openvino \
bash -c "/opt/intel/openvino/deployment_tools/model_optimizer/mo.py \
    --framework caffe \
    --input_model face_mask.caffemodel \
    --input_proto deploy.prototxt"
```

- Download face detection model from the model zoo, which will produce `.xml` and `.bin` files.
```bash
docker run --rm -ti \
--volume "$PWD":/app \
--env DISPLAY=$DISPLAY \
--volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
mmphego/intel-openvino \
bash -c "/opt/intel/openvino/deployment_tools/open_model_zoo/tools/downloader/downloader.py \
    --name face-detection-adas-0001 \
    --precision FP16"
```

## Usage

```bash
$ python main.py -h

usage: main.py [-h] -f FACE_MODEL -m MASK_MODEL -i INPUT [-d DEVICE]
               [--face_prob_threshold FACE_PROB_THRESHOLD]
               [--mask_prob_threshold MASK_PROB_THRESHOLD] [--enable-speech]
               [--tts TTS] [--ffmpeg] [--show-bbox] [--debug]

optional arguments:
  -h, --help            show this help message and exit
  -f FACE_MODEL, --face-model FACE_MODEL
                        Path to an xml file with a trained model.
  -m MASK_MODEL, --mask-model MASK_MODEL
                        Path to an xml file with a trained model.
  -i INPUT, --input INPUT
                        Path to image or video file or 'cam' for Webcam.
  -d DEVICE, --device DEVICE
                        Specify the target device to infer on: CPU, GPU, FPGA
                        or MYRIAD is acceptable. Sample will look for a
                        suitable plugin for device specified (CPU by default)
  --face_prob_threshold FACE_PROB_THRESHOLD
                        Probability threshold for face detections filtering
                        (Default: 0.8)
  --mask_prob_threshold MASK_PROB_THRESHOLD
                        Probability threshold for face mask detections
                        filtering(Default: 0.3)
  --enable-speech       Enable speech notification.
  --tts TTS             Text-to-Speech, used for notification.
  --ffmpeg              Flush video to FFMPEG.
  --show-bbox           Show bounding box and stats on screen [debugging].
  --debug               Show output on screen [debugging].

```

### Example Usage

```bash
make example
```

![ezgif-4-993045572ebb](https://user-images.githubusercontent.com/7910856/88237923-a8514500-cc80-11ea-9cc8-0692eb0c4d6e.gif)

### Packaging the Application
We can use the [Deployment Manager](https://docs.openvinotoolkit.org/latest/_docs_install_guides_deployment_manager_tool.html) present in OpenVINO to create a runtime package from our application. These packages can be easily sent to other hardware devices to be deployed.

To deploy the application to various devices using the Deployment Manager run the steps below.

Note: Choose from the devices listed below.

```bash
DEVICE='cpu' # or gpu, vpu, gna, hddl
docker run --rm -ti \
--volume "$PWD":/app \
mmphego/intel-openvino bash -c "\
  python /opt/intel/openvino/deployment_tools/tools/deployment_manager/deployment_manager.py \
  --targets ${DEVICE} \
  --user_data /app \
  --output_dir . \
  --archive_name face_mask_detection_${DEVICE}"

```

## Credit

- Face mask detection caffe model: [https://github.com/didi/maskdetection](https://github.com/didi/maskdetection)
- [COVID-19: Face Mask Detector with OpenCV, Keras/TensorFlow, and Deep Learning by Adrian Rosebrock ](https://www.pyimagesearch.com/2020/05/04/covid-19-face-mask-detector-with-opencv-keras-tensorflow-and-deep-learning/)


CHANGELOG.md
# Changelog

## [v1.1](https://github.com/mmphego/face_mask_detection_openvino/tree/v1.1) (2020-09-11)

[Full Changelog](https://github.com/mmphego/face_mask_detection_openvino/compare/v1.0...v1.1)

**Merged pull requests:**

- Complete rework of the code [\#3](https://github.com/mmphego/face_mask_detection_openvino/pull/3) ([mmphego](https://github.com/mmphego))

## [v1.0](https://github.com/mmphego/face_mask_detection_openvino/tree/v1.0) (2020-07-21)

[Full Changelog](https://github.com/mmphego/face_mask_detection_openvino/compare/81a8c4842bfec8e32d4b41d3a8067325be61e1f6...v1.0)

**Closed issues:**

- Multiple results [\#2](https://github.com/mmphego/face_mask_detection_openvino/issues/2)
- Unsupported layers found IECore OpenVINO MYRIAD [\#1](https://github.com/mmphego/face_mask_detection_openvino/issues/1)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


inference.py
import cv2
import numpy as np

from pyvino_utils.models.openvino_base.base_model import Base

__all__ = [
    "FaceDetection",
    "MaskDetection",
]


COLOR = {"Green": (0, 255, 0), "Red": (0, 0, 255)}


class FaceDetection(Base):
    """Class for the Face Detection Model."""

    def __init__(
        self,
        model_name,
        source_width=None,
        source_height=None,
        device="CPU",
        threshold=0.60,
        extensions=None,
        **kwargs,
    ):
        super().__init__(
            model_name,
            source_width,
            source_height,
            device,
            threshold,
            extensions,
            **kwargs,
        )

    def preprocess_output(self, inference_results, image, show_bbox=False, **kwargs):
        """Draw bounding boxes onto the Face Detection frame."""
        results = {}
        if not (self._init_image_w and self._init_image_h):
            raise RuntimeError("Initial image width and height cannot be None.")
        if len(inference_results) == 1:
            inference_results = inference_results[0]

        bbox_coord = []
        for box in inference_results[0][0]:  # Output shape is 1x1xNx7
            conf = box[2]
            if conf >= self.threshold:
                xmin = int(box[3] * self._init_image_w)
                ymin = int(box[4] * self._init_image_h)
                xmax = int(box[5] * self._init_image_w)
                ymax = int(box[6] * self._init_image_h)
                bbox_coord.append((xmin, ymin, xmax, ymax))
                if show_bbox:
                    self.draw_output(image, xmin, ymin, xmax, ymax, **kwargs)

        results["image"] = image
        results["bbox_coord"] = bbox_coord
        return results

    @staticmethod
    def draw_output(
        image,
        xmin,
        ymin,
        xmax,
        ymax,
        label="Person",
        padding_size=(0.05, 0.25),
        scale=2,
        thickness=2,
        **kwargs,
    ):
        _label = None
        if kwargs.get("mask_detected"):
            _label = (
                (f"{label} Wearing Mask", COLOR["Green"])
                if float(kwargs.get("mask_detected")) > kwargs.get("threshold", 0.1)
                else (f"{label} NOT wearing a Mask!!!", COLOR["Red"])
            )
            # print(_label)

        label = _label if _label is not None else (label, COLOR["Green"])

        cv2.rectangle(
            image, (xmin, ymin), (xmax, ymax,), color=label[1], thickness=thickness,
        )
        ((label_width, label_height), _) = cv2.getTextSize(
            label[0], cv2.FONT_HERSHEY_PLAIN, fontScale=scale, thickness=thickness,
        )

        cv2.putText(
            image,
            label[0],
            org=(image.shape[0] // 3, image.shape[1] // 3),
            fontFace=cv2.FONT_HERSHEY_PLAIN,
            fontScale=scale,
            color=label[1],
            thickness=thickness,
        )


class MaskDetection(Base):
    """Class for the Mask Detection Model."""

    def __init__(
        self,
        model_name,
        source_width=None,
        source_height=None,
        device="CPU",
        threshold=0.60,
        extensions=None,
        **kwargs,
    ):
        super().__init__(
            model_name,
            source_width,
            source_height,
            device,
            threshold,
            extensions,
            **kwargs,
        )

    def preprocess_output(self, inference_results, image, show_bbox=False, **kwargs):
        results = {}
        results["flattened_predictions"] = np.vstack(inference_results).ravel()
        results["image"] = image
        return results

    def draw_output(
        self, image, inference_results, **kwargs,
    ):
        pass


main.py
#!/usr/bin/env python3

from argparse import ArgumentParser

from responsive_voice.voices import UKEnglishMale

from inference import FaceDetection, MaskDetection
from pyvino_utils import InputFeeder


def arg_parser():
    """Parse command line arguments.

    :return: command line arguments
    """
    parser = ArgumentParser()
    parser.add_argument(
        "-f",
        "--face-model",
        required=True,
        type=str,
        help="Path to an xml file with a trained model.",
    )
    parser.add_argument(
        "-m",
        "--mask-model",
        required=True,
        type=str,
        help="Path to an xml file with a trained model.",
    )
    parser.add_argument(
        "-i",
        "--input",
        required=True,
        type=str,
        help="Path to image or video file or 'cam' for Webcam.",
    )

    parser.add_argument(
        "-d",
        "--device",
        default="CPU",
        help="Specify the target device to infer on: "
        "CPU, GPU, FPGA or MYRIAD is acceptable. Sample "
        "will look for a suitable plugin for device "
        "specified (CPU by default)",
    )
    parser.add_argument(
        "--face_prob_threshold",
        type=float,
        default=0.8,
        help="Probability threshold for face detections filtering (Default: 0.8)",
    )
    parser.add_argument(
        "--mask_prob_threshold",
        type=float,
        default=0.3,
        help="Probability threshold for face mask detections filtering" "(Default: 0.3)",
    )
    parser.add_argument(
        "--enable-speech", action="store_true", help="Enable speech notification.",
    )
    parser.add_argument(
        "--tts",
        type=str,
        default="Please wear your MASK!!",
        help="Text-to-Speech, used for notification.",
    )
    parser.add_argument(
        "--ffmpeg", action="store_true", help="Flush video to FFMPEG.",
    )
    parser.add_argument(
        "--show-bbox",
        action="store_true",
        help="Show bounding box and stats on screen [debugging].",
    )
    parser.add_argument(
        "--debug", action="store_true", help="Show output on screen [debugging].",
    )

    return parser.parse_args()


def main(args):
    """
    Initialize the inference network, stream video to network,
    and output stats and video.

    :param args: Command line arguments parsed by `build_argparser()`
    :param client: MQTT client
    :return: None
    """
    # Initialise the video stream
    input_feed = InputFeeder(input_feed=args.input)
    # Initialise the speech output
    if args.enable_speech:
        # TODO: Add args for selecting language, accent and male/female voice
        engine = UKEnglishMale()
        speak = engine.get_mp3(args.tts)
    # Initialise the class
    face_detection = FaceDetection(
        model_name=args.face_model,
        device=args.device,
        threshold=args.face_prob_threshold,
        input_feed=input_feed,
    )
    mask_detection = MaskDetection(
        model_name=args.mask_model,
        device=args.device,
        threshold=args.mask_prob_threshold,
    )

    count = 0
    face_detect_infer_time = 0
    mask_detect_infer_time = 0
    mask_detected_prob = -1
    try:
        # TODO: Convert to contextmanager
        for frame in input_feed.next_frame():
            count += 1

            fd_results = face_detection.predict(
                frame, show_bbox=args.show_bbox, mask_detected=mask_detected_prob
            )
            face_bboxes = fd_results["process_output"]["bbox_coord"]
            if face_bboxes:
                for face_bbox in face_bboxes:
                    # Useful resource:
                    # https://www.pyimagesearch.com/2018/09/24/opencv-face-recognition/

                    # Face bounding box coordinates cropped from the face detection
                    # inference are face_bboxes i.e `xmin, ymin, xmax, ymax`
                    # Therefore the face can be cropped by:
                    # frame[face_bbox[1]:face_bbox[3], face_bbox[0]:face_bbox[2]]

                    # extract the face ROI
                    (x, y, w, h) = face_bbox
                    face = frame[y:h, x:w]
                    (face_height, face_width) = face.shape[:2]
                    # Crop and show face
                    # input_feed.show(frame[y:h, x:w], "face")

                    # ensure the face width and height are sufficiently large
                    if face_height < 20 or face_width < 20:
                        continue

                    md_results = mask_detection.predict(
                        face, show_bbox=args.show_bbox, frame=frame
                    )
                    mask_detected_prob = md_results["process_output"][
                        "flattened_predictions"
                    ]
                    if (
                        int(count) % 200 == 1
                        and args.enable_speech
                        and float(mask_detected_prob) < args.mask_prob_threshold
                    ):
                        engine.play_mp3(speak)

            if args.debug:
                text = f"Face Detection Inference time: {face_detect_infer_time:.3f} ms"
                input_feed.add_text(text, frame, (15, input_feed.source_height - 80))
                text = (
                    f"Face Mask Detection Inference time: {mask_detect_infer_time:.3f} ms"
                )
                input_feed.add_text(text, frame, (15, input_feed.source_height - 60))

                input_feed.show(input_feed.resize(frame))

    finally:
        input_feed.close()


if __name__ == "__main__":
    # Grab command line args
    args = arg_parser()
    # Perform inference on the input stream
    main(args)
